lib/device-connections-factory.js

```diff
@@ -0,0 +1,178 @@
+import _ from 'lodash';
+import net from 'net';
+import B from 'bluebird';
+import { logger } from 'appium-support';
+import { utilities } from 'appium-ios-device';
+import { checkPortStatus } from 'portscanner';
+
+const log = logger.getLogger('iProxy');
+
+class iProxy {
+  constructor (udid, localport, deviceport) {
+    this.localport = parseInt(localport, 10);
+    this.deviceport = parseInt(deviceport, 10);
+    this.udid = udid;
+    this.serverSocket = null;
+  }
+
+  async start () {
+    if (this.serverSocket) {
+      return;
+    }
+    this.serverSocket = net.createServer(async (connection) => {
+      try {
+        const socket = await utilities.connectPort(this.udid, this.deviceport);
+        socket.on('close', connection.destroy);
+        socket.on('error', log.error);
+        connection.on('close', socket.destroy);
+        connection.on('error', log.error);
+        connection.pipe(socket);
+        socket.pipe(connection);
+      } catch (e) {
+        log.warn(e.message);
+        connection.destroy();
+      }
+    });
+    const status = new B((resolve, reject) => {
+      this.serverSocket.once('listening', resolve);
+      this.serverSocket.once('error', reject);
+    });
+    this.serverSocket.listen(this.localport);
+    await status;
+  }
+
+  quit () {
+    if (!this.serverSocket) {
+      return;
+    }
+    this.serverSocket.close();
+    this.serverSocket = null;
+  }
+}
+
+class DeviceConnectionsFactory {
+  constructor () {
+    this._connectionsMapping = {};
+  }
+
+  _toKey (udid = null, port = null) {
+    return `${udid || ''}:${port || ''}`;
+  }
+
+  _releaseProxiedConnections (connectionKeys) {
+    const result = [];
+    connectionKeys
+      .filter((k) => _.has(this._connectionsMapping[k], 'iproxy'))
+      .map((k) => {
+        log.info(`Releasing the listener for '${k}'`);
+        try {
+          this._connectionsMapping[k].iproxy.quit();
+        } catch (err) {
+          log.warn(`Cannot release the listener for '${k}': ${err.message}`);
+        }
+        result.push(k);
+      });
+    return result;
+  }
+
+  listConnections (udid = null, port = null, strict = false) {
```

Comment:
And

> the idea here is to cleanup all connections for the device under test after the session is closed, independently of the port

Can I clarify this behaviour?

- `releaseConnection` falls here as `restrict = true` in 
    - returns `[]` most case since https://github.com/appium/appium-xcuitest-driver/pull/1066/files#diff-e4bb14e6d9c2a2322696a7de176a736cR162 calls here as `8100, null, true`
- `releaseConnection` in lib/commands/recordscreen.js returns the pair
    - the argument is `8100, xxxxxxxx`

I thought cleaning up all ports happen in lib/commands/recordscreen.js
Is this opposite? Because `stop` cleanes all of connections and recordscreen.js cleans only particular connection

Reply:
> Because I assumed stop cleans all of the connections and recordscreen.js cleans only particular connection are expected behaviour.

Yes, this is the correct assumption.  I'll add unit tests to make the logic cleaner